# Q1. Write a Python function that takes in two arrays of data and calculates the F-value for a variance ratio test. The function should return the F-value and the corresponding p-value for the test.

In [1]:
import numpy as np
from scipy import stats

def variance_ratio_test(data1, data2):
    """
    Perform a variance ratio test (F-test) to compare the variances of two datasets.

    Parameters:
    - data1: Array of data points from the first sample.
    - data2: Array of data points from the second sample.

    Returns:
    - F-value: The calculated F-statistic.
    - p-value: The corresponding p-value for the test.
    """
    # Calculate the variances of the two samples
    var1 = np.var(data1, ddof=1)  # Sample variance for data1
    var2 = np.var(data2, ddof=1)  # Sample variance for data2

    # Calculate the F-statistic (larger variance / smaller variance)
    if var1 > var2:
        F_value = var1 / var2
    else:
        F_value = var2 / var1

    # Degrees of freedom
    df1 = len(data1) - 1
    df2 = len(data2) - 1

    # Calculate the p-value from the F-distribution
    p_value = 1 - stats.f.cdf(F_value, df1, df2)

    return F_value, p_value

# Example usage:
data1 = np.random.normal(0, 1, 100)  # Sample 1 with mean 0 and standard deviation 1
data2 = np.random.normal(0, 2, 100)  # Sample 2 with mean 0 and standard deviation 2

F_value, p_value = variance_ratio_test(data1, data2)
print(f"F-value: {F_value:.4f}, p-value: {p_value:.4f}")


F-value: 3.5957, p-value: 0.0000


# Q2. Given a significance level of 0.05 and the degrees of freedom for the numerator and denominator of an F-distribution, write a Python function that returns the critical F-value for a two-tailed test.

In [2]:
from scipy import stats

def critical_F_value(alpha, df1, df2):
    """
    Calculate the critical F-value for a two-tailed test at a given significance level.

    Parameters:
    - alpha: Significance level (e.g., 0.05).
    - df1: Degrees of freedom for the numerator.
    - df2: Degrees of freedom for the denominator.

    Returns:
    - critical_F: The critical F-value for the given significance level.
    """
    # Calculate the critical F-value for a two-tailed test
    # For a two-tailed test, we split the alpha level into two tails, so we divide alpha by 2
    critical_F = stats.f.ppf(1 - alpha/2, df1, df2)

    return critical_F

# Example usage:
alpha = 0.05  # Significance level
df1 = 5       # Degrees of freedom for the numerator
df2 = 10      # Degrees of freedom for the denominator

critical_F = critical_F_value(alpha, df1, df2)
print(f"Critical F-value: {critical_F:.4f}")


Critical F-value: 4.2361


# Q3. Write a Python program that generates random samples from two normal distributions with known variances and uses an F-test to determine if the variances are equal. The program should output the F- value, degrees of freedom, and p-value for the test.

In [3]:
import numpy as np
from scipy import stats

def f_test_for_variances(sample1, sample2):
    """
    Perform an F-test to compare the variances of two normal distributions.

    Parameters:
    - sample1: Array of data points from the first sample.
    - sample2: Array of data points from the second sample.

    Returns:
    - F-value: The calculated F-statistic.
    - df1: Degrees of freedom for sample1.
    - df2: Degrees of freedom for sample2.
    - p-value: The corresponding p-value for the F-test.
    """
    # Calculate sample variances
    var1 = np.var(sample1, ddof=1)
    var2 = np.var(sample2, ddof=1)

    # Calculate the F-statistic (larger variance / smaller variance)
    if var1 > var2:
        F_value = var1 / var2
    else:
        F_value = var2 / var1

    # Degrees of freedom
    df1 = len(sample1) - 1  # Degrees of freedom for sample1
    df2 = len(sample2) - 1  # Degrees of freedom for sample2

    # Calculate the p-value from the F-distribution
    p_value = 1 - stats.f.cdf(F_value, df1, df2)

    return F_value, df1, df2, p_value

# Generate random samples from two normal distributions
np.random.seed(42)
sample1 = np.random.normal(loc=0, scale=5, size=100)  # Sample 1 (mean=0, std=5)
sample2 = np.random.normal(loc=0, scale=10, size=100)  # Sample 2 (mean=0, std=10)

# Perform the F-test for equal variances
F_value, df1, df2, p_value = f_test_for_variances(sample1, sample2)

# Output the results
print(f"F-value: {F_value:.4f}")
print(f"Degrees of Freedom (Sample 1): {df1}")
print(f"Degrees of Freedom (Sample 2): {df2}")
print(f"P-value: {p_value:.4f}")

# Interpretation
if p_value < 0.05:
    print("Reject the null hypothesis: The variances are significantly different.")
else:
    print("Fail to reject the null hypothesis: The variances are equal.")


F-value: 4.4109
Degrees of Freedom (Sample 1): 99
Degrees of Freedom (Sample 2): 99
P-value: 0.0000
Reject the null hypothesis: The variances are significantly different.


# Q4.The variances of two populations are known to be 10 and 15. A sample of 12 observations is taken from each population. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

## Step 1: Define the Hypotheses
- **Null Hypothesis (\(H_0\))**: The variances of the two populations are equal, i.e., \(\sigma_1^2 = \sigma_2^2\).
- **Alternative Hypothesis (\(H_a\))**: The variances of the two populations are not equal, i.e., \(\sigma_1^2 \neq \sigma_2^2\).

## Step 2: Calculate the F-statistic
The F-statistic is the ratio of the two variances, where the larger variance is placed in the numerator:
\[
F = \frac{s_1^2}{s_2^2}
\]
where:
- \(s_1^2 = 10\) is the variance of the first population,
- \(s_2^2 = 15\) is the variance of the second population.

Thus, the F-statistic is:
\[
F = \frac{15}{10} = 1.5
\]

## Step 3: Degrees of Freedom
Since the sample sizes are both 12, the degrees of freedom for each sample are:
- \(df_1 = 12 - 1 = 11\),
- \(df_2 = 12 - 1 = 11\).

## Step 4: Find the Critical F-values
At a 5% significance level (\(\alpha = 0.05\)), for a two-tailed test, we need to find the critical F-values using the degrees of freedom for both samples.

## Step 5: Decision Rule
- If the calculated F-statistic is less than the lower critical value or greater than the upper critical value, we **reject** the null hypothesis.
- If the calculated F-statistic lies within the critical range, we **fail to reject** the null hypothesis.


#Q5. A manufacturer claims that the variance of the diameter of a certain product is 0.005. A sample of 25 products is taken, and the sample variance is found to be 0.006. Conduct an F-test at the 1% significance level to determine if the claim is justified.

## Step 1: Define the Hypotheses
- **Null Hypothesis (\(H_0\))**: The variance of the product diameter is 0.005, i.e., \(\sigma^2 = 0.005\).
- **Alternative Hypothesis (\(H_a\))**: The variance of the product diameter is not 0.005, i.e., \(\sigma^2 \neq 0.005\).

## Step 2: Calculate the F-statistic
The F-statistic is calculated as the ratio of the sample variance to the claimed variance:
\[
F = \frac{s^2}{\sigma^2}
\]
where:
- \(s^2 = 0.006\) is the sample variance,
- \(\sigma^2 = 0.005\) is the claimed variance.

Thus, the F-statistic is:
\[
F = \frac{0.006}{0.005} = 1.2
\]

## Step 3: Degrees of Freedom
For this test, the degrees of freedom for the sample variance are calculated as:
\[
df = n - 1 = 25 - 1 = 24
\]
where \(n = 25\) is the sample size.

## Step 4: Find the Critical F-values
At a 1% significance level (\(\alpha = 0.01\)), for a two-tailed test with \(df_1 = 24\) and \(df_2 = 24\), we need to find the critical F-values.

Since it's a two-tailed test, we divide the significance level by 2, which gives us \(\alpha/2 = 0.005\) for each tail.

## Step 5: Decision Rule
- If the calculated F-statistic is less than the lower critical value or greater than the upper critical value, we **reject** the null hypothesis.
- If the calculated F-statistic lies within the critical range, we **fail to reject** the null hypothesis.


# Q6. Write a Python function that takes in the degrees of freedom for the numerator and denominator of an F-distribution and calculates the mean and variance of the distribution. The function should return the mean and variance as a tuple.

In [6]:
def f_distribution_mean_variance(df1, df2):
    """
    Calculate the mean and variance of an F-distribution given the degrees of freedom for the numerator (df1) and the denominator (df2).

    Args:
    df1 (int): Degrees of freedom for the numerator.
    df2 (int): Degrees of freedom for the denominator.

    Returns:
    tuple: (mean, variance) of the F-distribution.
    """
    if df2 <= 2:
        raise ValueError("Degrees of freedom for the denominator (df2) must be greater than 2 to calculate the mean.")

    mean = df2 / (df2 - 2)

    if df2 <= 4:
        raise ValueError("Degrees of freedom for the denominator (df2) must be greater than 4 to calculate the variance.")

    variance = (2 * (df2 ** 2) * (df1 + df2 - 2)) / (df1 * (df2 - 2) ** 2 * (df2 - 4))

    return mean, variance

# Example usage
df1 = 5  # Degrees of freedom for the numerator
df2 = 10  # Degrees of freedom for the denominator

mean, variance = f_distribution_mean_variance(df1, df2)
print(f"Mean: {mean:.4f}")
print(f"Variance: {variance:.4f}")


Mean: 1.2500
Variance: 1.3542


#Q7. A random sample of 10 measurements is taken from a normal population with unknown variance. The sample variance is found to be 25. Another random sample of 15 measurements is taken from another normal population with unknown variance, and the sample variance is found to be 20. Conduct an F-test at the 10% significance level to determine if the variances are significantly different.

## Hypotheses
- Null Hypothesis (\(H_0\)): The variances of the two populations are equal, i.e., \(\sigma_1^2 = \sigma_2^2\).
- Alternative Hypothesis (\(H_a\)): The variances of the two populations are not equal, i.e., \(\sigma_1^2 \neq \sigma_2^2\).

## Step 1: F-statistic Calculation
The F-statistic is calculated as the ratio of the two sample variances:
\[
F = \frac{s_1^2}{s_2^2}
\]
where:
- \(s_1^2 = 25\) (variance from the first sample),
- \(s_2^2 = 20\) (variance from the second sample).

Thus, the F-statistic is:
\[
F = \frac{25}{20} = 1.25
\]

## Step 2: Degrees of Freedom
For each sample, the degrees of freedom are:
- \(df_1 = n_1 - 1 = 10 - 1 = 9\),
- \(df_2 = n_2 - 1 = 15 - 1 = 14\).

## Step 3: Critical F-values
The critical F-values for a two-tailed test at a 10% significance level (\(\alpha = 0.10\)) with degrees of freedom \(df_1 = 9\) and \(df_2 = 14\) are:
- Lower critical value: \( F_{\text{low}} = 0.33 \)
- Upper critical value: \( F_{\text{high}} = 2.65 \)

## Step 4: Decision Rule
If the calculated F-statistic falls outside the range defined by the critical F-values, we reject the null hypothesis. Otherwise, we fail to reject it.

## Step 5: Conclusion
The calculated F-statistic is:
\[
F = 1.25
\]
This value lies within the range of \(0.33 \leq F \leq 2.65\), so we **fail to reject the null hypothesis**.

### Final Conclusion:
There is not enough evidence at the 10% significance level to conclude that the variances of the two populations are significantly different.


#Q8. The following data represent the waiting times in minutes at two different restaurants on a Saturday night: Restaurant A: 24, 25, 28, 23, 22, 20, 27; Restaurant B: 31, 33, 35, 30, 32, 36. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

## Hypotheses
- Null Hypothesis (\(H_0\)): The variances of the waiting times at both restaurants are equal, i.e., \(\sigma_1^2 = \sigma_2^2\).
- Alternative Hypothesis (\(H_a\)): The variances of the waiting times at both restaurants are not equal, i.e., \(\sigma_1^2 \neq \sigma_2^2\).

## Step 1: Sample Variances
- Sample data for Restaurant A: 24, 25, 28, 23, 22, 20, 27
- Sample data for Restaurant B: 31, 33, 35, 30, 32, 36

The sample variances are:
- Sample variance for Restaurant A: \(s_A^2 = 7.81\)
- Sample variance for Restaurant B: \(s_B^2 = 5.37\)

## Step 2: F-statistic Calculation
The F-statistic is calculated as:
\[
F = \frac{s_A^2}{s_B^2} = \frac{7.81}{5.37} = 1.46
\]

## Step 3: Degrees of Freedom
- Degrees of freedom for Restaurant A: \(df_1 = 7 - 1 = 6\)
- Degrees of freedom for Restaurant B: \(df_2 = 6 - 1 = 5\)

## Step 4: Critical F-values
At a 5% significance level (\(\alpha = 0.05\)) for a two-tailed test, the critical F-values are:
- Lower critical value: \( F_{\text{low}} = 0.17 \)
- Upper critical value: \( F_{\text{high}} = 6.98 \)

## Step 5: Decision Rule
If the calculated F-statistic falls outside the range of \(F_{\text{low}}\) and \(F_{\text{high}}\), we reject the null hypothesis. Otherwise, we fail to reject it.

## Step 6: Conclusion
The calculated F-statistic is:
\[
F = 1.46
\]
This value lies within the range of \(0.17 \leq F \leq 6.98\), so we **fail to reject the null hypothesis**.

### Final Conclusion:
There is not enough evidence at the 5% significance level to conclude that the variances of the waiting times at Restaurant A and Restaurant B are significantly different.


# Q9. The following data represent the test scores of two groups of students: Group A: 80, 85, 90, 92, 87, 83; Group B: 75, 78, 82, 79, 81, 84. Conduct an F-test at the 1% significance level to determine if the variances are significantly different.

## Hypotheses
- Null Hypothesis (\(H_0\)): The variances of the test scores in both groups are equal, i.e., \(\sigma_1^2 = \sigma_2^2\).
- Alternative Hypothesis (\(H_a\)): The variances of the test scores in both groups are not equal, i.e., \(\sigma_1^2 \neq \sigma_2^2\).

## Step 1: Sample Variances
- Sample data for Group A: 80, 85, 90, 92, 87, 83
- Sample data for Group B: 75, 78, 82, 79, 81, 84

The sample variances are:
- Sample variance for Group A: \(s_A^2 = 19.77\)
- Sample variance for Group B: \(s_B^2 = 10.17\)

## Step 2: F-statistic Calculation
The F-statistic is calculated as:
\[
F = \frac{s_A^2}{s_B^2} = \frac{19.77}{10.17} = 1.94
\]

## Step 3: Degrees of Freedom
- Degrees of freedom for Group A: \(df_1 = 6 - 1 = 5\)
- Degrees of freedom for Group B: \(df_2 = 6 - 1 = 5\)

## Step 4: Critical F-values
At a 1% significance level (\(\alpha = 0.01\)) for a two-tailed test, the critical F-values are:
- Lower critical value: \( F_{\text{low}} = 0.067 \)
- Upper critical value: \( F_{\text{high}} = 14.94 \)

## Step 5: Decision Rule
If the calculated F-statistic falls outside the range of \(F_{\text{low}}\) and \(F_{\text{high}}\), we reject the null hypothesis. Otherwise, we fail to reject it.

## Step 6: Conclusion
The calculated F-statistic is:
\[
F = 1.94
\]
This value lies within the range of \(0.067 \leq F \leq 14.94\), so we **fail to reject the null hypothesis**.

### Final Conclusion:
There is not enough evidence at the 1% significance level to conclude that the variances of the test scores in Group A and Group B are significantly different.
